### PR TITLE
feat(controller): Sum resources duration for DAGs and steps

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Related to #3864

### CODE REVIEW

**Comment by alexec** [comment]
@JPZ13  can you try `argoproj/workflow-controller:res-dur` please?

File: workflow/graph/topsort.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+// Return a list of node IDs that can be iterated over know we'll visit each one only once and that
+// we'll visit downstream dependencies before upstream ones.
+// Children can be:
+// * "missing" - appear in `node.Children` but missing from `nodes` - callers must cope with this and probably skip them
+func TopSort(nodes wfv1.Nodes) ([]string, error) {
+	g, err := graph(nodes)
```
Comment (alexec) [inline workflow/graph/topsort.go:14]: @simster7 pursuant to our chat, this shows the API I'd like. I'm not really fussed about the implementation (as long as it works).

**Comment by JPZ13** [comment]
@alexec Will do

**Comment by JPZ13** [comment]
@alexec Code looks really good - my only question is about a nested case. I'm not seeing the parent resources duration be the sum of the children. Is that expected? Here's my output for the `./tests/e2e/functional/dag-nested.yaml` run:
```
Name:                dag-nested-4xx2m
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Conditions:
 Completed           True
Created:             Mon Sep 21 16:28:48 -0400 (2 minutes ago)
Started:             Mon Sep 21 16:28:48 -0400 (2 minutes ago)
Finished:            Mon Sep 21 16:29:52 -0400 (1 minute ago)
Duration:            1 minute 4 seconds
ResourcesDuration:   2s*(1 cpu),1s*(100Mi memory)

STEP                 TEMPLATE        PODNAME                      DURATION  ARTIFACTS  MESSAGE  RESOURCESDURATION             NODENAME
 ✔ dag-nested-4xx2m  diamond                                                                    1s*(1 cpu),0s*(100Mi memory)
 ├-✔ A               nested-diamond                                                             1s*(1 cpu),0s*(100Mi memory)
 | ├-✔ A             echo            dag-nested-4xx2m-2908879225  3s        main-logs           1s*(1 cpu),0s*(100Mi memory)  docker-desktop
 | ├-✔ B             echo            dag-nested-4xx2m-2858546368  6s        main-logs           3s*(1 cpu),1s*(100Mi memory)  docker-desktop
 | ├-✔ C             echo            dag-nested-4xx2m-2875323987  6s        main-logs           3s*(1 cpu),1s*(100Mi memory)  docker-desktop
 | └-✔ D             echo            dag-nested-4xx2m-2959212082  3s        main-logs           2s*(1 cpu),1s*(100Mi memory)  docker-desktop
 ├-✔ B               nested-diamond                                                             2s*(1 cpu),1s*(100Mi memory)
 | ├-✔ A             echo            dag-nested-4xx2m-1211952452  5s        main-logs           1s*(100Mi memory),2s*(1 cpu)  docker-desktop
 | ├-✔ B             echo            dag-nested-4xx2m-1262285309  10s       main-logs           5s*(1 cpu),3s*(100Mi memory)  docker-desktop
 | ├-✔ C             echo            dag-nested-4xx2m-1245507690  9s        main-logs           4s*(1 cpu),2s*(100Mi memory)  docker-desktop
 | └-✔ D             echo            dag-nested-4xx2m-1161619595  6s        main-logs           2s*(1 cpu),1s*(100Mi memory)  docker-desktop
 ├-✔ C               nested-diamond                                                             3s*(1 cpu),1s*(100Mi memory)
 | ├-✔ A             echo            dag-nested-4xx2m-3799189731  6s        main-logs           3s*(1 cpu),1s*(100Mi memory)  docker-desktop
 | ├-✔ B             echo            dag-nested-4xx2m-3815967350  9s        main-logs           4s*(1 cpu),2s*(100Mi memory)  docker-desktop
 | ├-✔ C             echo            dag-nested-4xx2m-3832744969  10s       main-logs           4s*(1 cpu),2s*(100Mi memory)  docker-desktop
 | └-✔ D             echo            dag-nested-4xx2m-3849522588  7s        main-logs           1s*(100Mi memory),3s*(1 cpu)  docker-desktop
 └-✔ D               nested-diamond                                                             1s*(1 cpu),0s*(100Mi memory)
   ├-✔ A             echo            dag-nested-4xx2m-4249798190  3s        main-logs           1s*(1 cpu),0s*(100Mi memory)  docker-desktop
   ├-✔ B             echo            dag-nested-4xx2m-4233020571  6s        main-logs           4s*(1 cpu),2s*(100Mi memory)  docker-desktop
   ├-✔ C             echo            dag-nested-4xx2m-4216242952  5s        main-logs           3s*(1 cpu),2s*(100Mi memory)  docker-desktop
   └-✔ D             echo            dag-nested-4xx2m-38718989    4s        main-logs           2s*(1 cpu),1s*(100Mi memory)  docker-desktop
```

**Comment by alexec** [comment]
> I'm not seeing the parent resources duration be the sum of the children. Is that expected?

Not at all. That is a bug. I'll investigate.

**Comment by alexec** [comment]
It is a zero-day bug.

**Comment by JPZ13** [comment]
Damn. Is there anything I can do to be helpful with on that? I don't have the context of the bug or nearly the same depth of knowledge of the codebase as you, but I'm happy to help run test cases or do anything you need for the next couple hours

**Comment by alexec** [comment]
> It is a zero-day bug.

I'm actually not sure it is a bug now. A misunderstand of how something works.

**Comment by alexec** [comment]
Algo v2. It was a very interesting challenge iterating over the graph and summing up child nodes. @JPZ13 , this is not actually what I expected (the CLI layout is very different to the UI), so it makes the values look wrong.

```
Name:                dag-nested-bm4zp
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Conditions:          
 Completed           True
Created:             Tue Sep 22 08:12:36 -0700 (4 minutes ago)
Started:             Tue Sep 22 08:12:36 -0700 (4 minutes ago)
Finished:            Tue Sep 22 08:13:14 -0700 (3 minutes ago)
Duration:            38 seconds
ResourcesDuration:   23s*(1 cpu),7s*(100Mi memory)

STEP                 TEMPLATE        PODNAME                      DURATION  ARTIFACTS  MESSAGE  RESOURCESDURATION              NODENAME
 ✔ dag-nested-bm4zp  diamond                                                                    23s*(1 cpu),7s*(100Mi memory)  
 ├-✔ A               nested-diamond                                                             23s*(1 cpu),7s*(100Mi memory)  
 | ├-✔ A             echo            dag-nested-bm4zp-2090476449  2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | ├-✔ B             echo            dag-nested-bm4zp-2040143592  1s        main-logs           0s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | ├-✔ C             echo            dag-nested-bm4zp-2056921211  2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | └-✔ D             echo            dag-nested-bm4zp-2140809306  1s        main-logs           0s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 ├-✔ B               nested-diamond                                                             17s*(1 cpu),7s*(100Mi memory)  
 | ├-✔ A             echo            dag-nested-bm4zp-1462094412  2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | ├-✔ B             echo            dag-nested-bm4zp-1512427269  2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | ├-✔ C             echo            dag-nested-bm4zp-1495649650  2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | └-✔ D             echo            dag-nested-bm4zp-1411761555  2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 ├-✔ C               nested-diamond                                                             17s*(1 cpu),7s*(100Mi memory)  
 | ├-✔ A             echo            dag-nested-bm4zp-854040075   2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | ├-✔ B             echo            dag-nested-bm4zp-870817694   2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | ├-✔ C             echo            dag-nested-bm4zp-887595313   2s        main-logs           1s*(1 cpu),0s*(100Mi memory)   k3d-k3s-default-server
 | └-✔ D             echo            dag-nested-bm4zp-904372932   2s        main-logs           0s*(100Mi memory),1s*(1 cpu)   k3d-k3s-default-server
 └-✔ D               nested-diamond                                                             13s*(1 cpu),7s*(100Mi memory)  
   ├-✔ A             echo            dag-nested-bm4zp-2373193270  8s        main-logs           7s*(1 cpu),4s*(100Mi memory)   k3d-k3s-default-server
   ├-✔ B             echo            dag-nested-bm4zp-2356415651  3s        main-logs           2s*(1 cpu),1s*(100Mi memory)   k3d-k3s-default-server
   ├-✔ C             echo            dag-nested-bm4zp-2339638032  3s        main-logs           2s*(1 cpu),1s*(100Mi memory)   k3d-k3s-default-server
   └-✔ D             echo            dag-nested-bm4zp-2457081365  2s        main-logs           2s*(1 cpu),1s*(100Mi memory)   k3d-k3s-default-server
```

image

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -1208,10 +1208,14 @@ func (in ResourceDuration) String() string {
 type ResourcesDuration map[apiv1.ResourceName]ResourceDuration
 
 func (in ResourcesDuration) Add(o ResourcesDuration) ResourcesDuration {
+	res := ResourcesDuration{}
```
Comment (JPZ13) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1211]: Good idea modifying this to make a deep copy

File: util/resource/updater.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+			wf.Status.ResourcesDuration = wf.Status.ResourcesDuration.Add(node.ResourcesDuration)
+		} else if node.Fulfilled() {
+			// compute the sum of all children
+			node.ResourcesDuration = resourceDuration(wf, node, make(map[string]bool))
+			wf.Status.Nodes[nodeID] = node
+		}
+	}
+}
+
+func resourceDuration(wf *wfv1.Workflow, node wfv1.NodeStatus, visited map[string]bool) wfv1.ResourcesDuration {
```
Comment (JPZ13) [inline util/resource/updater.go:22]: Love seeing dynamic programming in the wild

**Comment by JPZ13** [comment]
@alexec Makes sense - the UI view definitely makes it clearer that the resources duration for the A DAG should contain the sums of the B, C, and D DAGs as well

**Comment by alexec** [comment]
Re-run jobs.

**Review by simster7 — APPROVED** [review]

**Pull request merged**